### ISSUE TITLE
AzureProxy paging operations

### ISSUE THREAD

**Issue by RikkiGibson** [Azure/autorest-clientruntime-for-java#222]
Right now we have special behavior for paging operations. The gist is that we have a simple `listSinglePage` variant, and an Observable variant which wraps the idea of getting a single page and recursively uses the next page link to get the next page.

It would probably cut down on our gen code a bit to have some special case for when `Observable<Page<Entity>>` is specified as a return type in AzureProxy where it would handle getting the nextPageLink out and performing the request for each successive page. Does that seem justified?

**Comment by daschult** [comment]
That seems very reasonable.

**Comment by RikkiGibson** [comment]
I want to make sure that cases like a user calling `getPagesAsync().first()` result in only the single necessary HTTP request being made. It looks like it's just necessary to make proper use of Observable.defer.
```
@Test
public void obsSemanticsTest() {
    Service14 service = createService(Service14.class);
    recursiveHelper(service).first().toBlocking().subscribe(new Action1<HttpBinJSON>() {
        @Override
        public void call(HttpBinJSON httpBinJSON) {
            System.out.println("I finished!");
        }
    });
}

private Observable<HttpBinJSON> recursiveHelper(final Service14 service) {
    return service.getAsync().toObservable().concatMap(new Func1<HttpBinJSON, Observable<? extends HttpBinJSON>>() {
        @Override
        public Observable<HttpBinJSON> call(HttpBinJSON httpBinJSON) {
            return Observable.just(httpBinJSON).concatWith(Observable.defer(new Func0<Observable<HttpBinJSON>>() {
                @Override
                public Observable<HttpBinJSON> call() {
                    return recursiveHelper(service);
                }
            }));
        }
    });
}
```
This example causes 1 HTTP request to be made. Leaving out the defer and calling recursiveHelper directly seemed to result in 2 HTTP requests being made.

**Comment by RikkiGibson** [comment]
Some of the stuff going on in here is pretty weird-- not certain about what the lazy behaviors of RxNettyAdapter are right now. Might follow up on this later.

**Comment by anuchandy** [comment]
Trying to understand, you mean something like this?

```java
public Single<Page> getSinglePageAsync(String nextLink) {
    return Single.defer(() -> {
        System.out.println("Doing n/w call for nextLink:" + nextLink);
        return Single.just(true).delay(1, TimeUnit.SECONDS);    // simulate network call delay
    })
    .map(b -> createPage(nextLink == "0" ? "1" : nextLink));
}

public Observable<Page> listAsync() {
    return getAllPages("0");
}

private Observable<Page> getAllPages(String nextLink) {
    if (nextLink == null) {
        Observable.empty();
    }
    return getSinglePageAsync(nextLink)
            .toObservable()
            .flatMap(page -> Observable.just(page).concatWith(getAllPages(page.nextLink)));
}
```

```java
class Page {
    public List<String> items;
    public String nextLink;
}

Page createPage(String nextLink) {
    Page page = new Page();
    page.items = new ArrayList<>();
    for (int c = 0; c < 5; c++) {
        page.items.add("page_" + nextLink + "_item_" + c);
    }
    page.nextLink = Integer.toString(Integer.parseInt(nextLink) + 1);
    if (page.nextLink == "10") {
        page.nextLink = null; // Last page
    }
    return page;
}
```

```java
@Test
public void testListAllPages() throws Exception {
    listAsync()
            .map(page -> page.items)
            .subscribe(System.out::println);
    Thread.sleep(12 * 1000);
}

@Test
public void testListFirstPage() throws Exception {
    listAsync()
            .take(1)
            .map(page -> page.items)
            .subscribe(System.out::println);
    Thread.sleep(5 * 1000);
}
```
output of first test
```
Doing n/w call for nextLink:0
[page_1_item_0, page_1_item_1, page_1_item_2, page_1_item_3, page_1_item_4]
Doing n/w call for nextLink:2
[page_2_item_0, page_2_item_1, page_2_item_2, page_2_item_3, page_2_item_4]
Doing n/w call for nextLink:3
[page_3_item_0, page_3_item_1, page_3_item_2, page_3_item_3, page_3_item_4]
Doing n/w call for nextLink:4
[page_4_item_0, page_4_item_1, page_4_item_2, page_4_item_3, page_4_item_4]
Doing n/w call for nextLink:5
[page_5_item_0, page_5_item_1, page_5_item_2, page_5_item_3, page_5_item_4]
Doing n/w call for nextLink:6
[page_6_item_0, page_6_item_1, page_6_item_2, page_6_item_3, page_6_item_4]
Doing n/w call for nextLink:7
[page_7_item_0, page_7_item_1, page_7_item_2, page_7_item_3, page_7_item_4]
Doing n/w call for nextLink:8
[page_8_item_0, page_8_item_1, page_8_item_2, page_8_item_3, page_8_item_4]
Doing n/w call for nextLink:9
[page_9_item_0, page_9_item_1, page_9_item_2, page_9_item_3, page_9_item_4]
Doing n/w call for nextLink:10
[page_10_item_0, page_10_item_1, page_10_item_2, page_10_item_3, page_10_item_4]
Doing n/w call for nextLink:11
[page_11_item_0, page_11_item_1, page_11_item_2, page_11_item_3, page_11_item_4]
Doing n/w call for nextLink:12
```

output of second test

```
Doing n/w call for nextLink:0
[page_1_item_0, page_1_item_1, page_1_item_2, page_1_item_3, page_1_item_4]
```

**Comment by anuchandy** [comment]
In your example, if leaving out defer cause two http call then that means `service.getAsync()` is not completely lazy, it runs as soon as we call it i.e. it is not waiting for a subscription (hence hot). May be we want to ensure rest proxy explicitly warp sendAsync calls in defer or implementation of sendAsync should do defer.

**Comment by RikkiGibson** [comment]
I think the best thing will be to tweak RxNettyAdapter so that calls are lazy as expected. It's strange because looking at the source it doesn't seem like it should be hot. OkHttpAdapter is cold.

**Comment by RikkiGibson** [comment]
For now it looks like we're probably going to just keep using `getSinglePageAsync` and `getNextPageAsync` methods and building everything out from that.

**Comment by RikkiGibson** [comment]
Implemented in #244
